akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -441,3 +445,102 @@ private[akka] final class ReducerState[T, R](val collector: java.util.stream.Col
   def finish(): R = collector.finisher().apply(reduced)
 }
 
+/**
+ * INTERNAL API
+ */
+final private[stream] class LazySink[T, M](sinkFactory: T ⇒ Future[Sink[T, M]], zeroMat: () ⇒ M) extends GraphStageWithMaterializedValue[SinkShape[T], Future[M]] {
```

Comment:
It seems like this approach creates a lot of code to maintain for each of the `lazy` Sink/Source/*, wouldn't laziness be worthwhile to implement at the AST level instead of the GraphStage level?